### PR TITLE
[Merged by Bors] - feat: port Order.Filter.Pi

### PR DESCRIPTION
**Pull request by urkud**
---

- [x] depends on: #1791

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1791
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1791~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Order/Filter/Pi.lean [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+end Pi
+
+/-! ### `n`-ary coproducts of filters -/
+
+section CoprodCat
+
+/-- Coproduct of filters. -/
+protected def Coprod (f : ∀ i, Filter (α i)) : Filter (∀ i, α i) :=
+  ⨆ i : ι, comap (eval i) (f i)
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Pi.lean:199]: I think this should become `coprod` + `subscript i`, just like `supr` became `sup_i` etc...

Comment (urkud) [reply Mathlib/Order/Filter/Pi.lean]: Renamed.

File: Mathlib/Order/Filter/Pi.lean [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  simp only [le_def, mem_map, mem_coprodᵢ_iff]
+  intro s h i
+  obtain ⟨t, H, hH⟩ := h i
+  exact ⟨{ x : α i | m i x ∈ t }, H, fun x hx => hH hx⟩
+#align filter.map_pi_map_Coprod_le Filter.map_pi_map_coprodᵢ_le
+
+theorem Tendsto.pi_map_coprod {g : ∀ i, Filter (β i)} (h : ∀ i, Tendsto (m i) (f i) (g i)) :
+    Tendsto (fun k : ∀ i, α i => fun i => m i (k i)) (Filter.coprodᵢ f) (Filter.coprodᵢ g) :=
+  map_pi_map_coprodᵢ_le.trans (coprodᵢ_mono h)
+#align filter.tendsto.pi_map_Coprod Filter.Tendsto.pi_map_coprod
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Pi.lean:269]:
```suggestion
theorem Tendsto.pi_map_coprodᵢ {g : ∀ i, Filter (β i)} (h : ∀ i, Tendsto (m i) (f i) (g i)) :
    Tendsto (fun k : ∀ i, α i => fun i => m i (k i)) (Filter.coprodᵢ f) (Filter.coprodᵢ g) :=
  map_pi_map_coprodᵢ_le.trans (coprodᵢ_mono h)
#align filter.tendsto.pi_map_Coprod Filter.Tendsto.pi_map_coprodᵢ
```

File: Mathlib/Order/Filter/Pi.lean [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+section CoprodCat
+
+/-- Coproduct of filters. -/
+protected def coprodᵢ (f : ∀ i, Filter (α i)) : Filter (∀ i, α i) :=
+  ⨆ i : ι, comap (eval i) (f i)
+#align filter.Coprod Filter.coprodᵢ
+
+theorem mem_coprodᵢ_iff {s : Set (∀ i, α i)} :
+    s ∈ Filter.coprodᵢ f ↔ ∀ i : ι, ∃ t₁ ∈ f i, eval i ⁻¹' t₁ ⊆ s := by simp [Filter.coprodᵢ]
+#align filter.mem_Coprod_iff Filter.mem_coprod_iff
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Pi.lean:204]:
```suggestion
#align filter.mem_Coprod_iff Filter.mem_coprodᵢ_iff
```

File: Mathlib/Order/Filter/Pi.lean [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+theorem compl_mem_coprodᵢ {s : Set (∀ i, α i)} :
+    sᶜ ∈ Filter.coprodᵢ f ↔ ∀ i, (eval i '' s)ᶜ ∈ f i :=
+  by simp only [Filter.coprodᵢ, mem_supᵢ, compl_mem_comap]
+#align filter.compl_mem_Coprod Filter.compl_mem_coprod
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Pi.lean:209]:
```suggestion
#align filter.compl_mem_Coprod Filter.compl_mem_coprodᵢ
```

File: Mathlib/Order/Filter/Pi.lean [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+theorem coprodᵢ_neBot_iff' :
+    NeBot (Filter.coprodᵢ f) ↔ (∀ i, Nonempty (α i)) ∧ ∃ d, NeBot (f d) := by
+  simp only [Filter.coprodᵢ, supᵢ_neBot, ← exists_and_left, ← comap_eval_neBot_iff']
+#align filter.Coprod_ne_bot_iff' Filter.coprodᵢ_neBot_iff'
+
+@[simp]
+theorem coprodᵢ_neBot_iff [∀ i, Nonempty (α i)] : NeBot (Filter.coprodᵢ f) ↔ ∃ d, NeBot (f d) := by
+  simp [coprodᵢ_neBot_iff', *]
+#align filter.Coprod_ne_bot_iff Filter.coprod_neBot_iff
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Pi.lean:219]:
```suggestion
#align filter.Coprod_ne_bot_iff Filter.coprodᵢ_neBot_iff
```

File: Mathlib/Order/Filter/Pi.lean [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+#align filter.Coprod_eq_bot_iff Filter.coprodᵢ_eq_bot_iff
+
+@[simp] theorem coprodᵢ_bot' : Filter.coprodᵢ (⊥ : ∀ i, Filter (α i)) = ⊥ :=
+  coprodᵢ_eq_bot_iff'.2 (Or.inr rfl)
+#align filter.Coprod_bot' Filter.coprodᵢ_bot'
+
+@[simp]
+theorem coprodᵢ_bot : Filter.coprodᵢ (fun _ => ⊥ : ∀ i, Filter (α i)) = ⊥ :=
+  coprodᵢ_bot'
+#align filter.Coprod_bot Filter.coprod_bot
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Pi.lean:238]:
```suggestion
#align filter.Coprod_bot Filter.coprodᵢ_bot
```

File: Mathlib/Order/Filter/Pi.lean [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+@[instance]
+theorem coprodᵢ_neBot [∀ i, Nonempty (α i)] [Nonempty ι] (f : ∀ i, Filter (α i))
+    [H : ∀ i, NeBot (f i)] : NeBot (Filter.coprodᵢ f) :=
+  (H (Classical.arbitrary ι)).coprodᵢ
+#align filter.Coprod_ne_bot Filter.coprodᵢ_neBot
+
+-- porting note: restore @[mono]
+theorem coprodᵢ_mono (hf : ∀ i, f₁ i ≤ f₂ i) : Filter.coprodᵢ f₁ ≤ Filter.coprodᵢ f₂ :=
+  supᵢ_mono fun i => comap_mono (hf i)
+#align filter.Coprod_mono Filter.coprod_mono
```
Comment (jcommelin) [inline Mathlib/Order/Filter/Pi.lean:253]:
```suggestion
#align filter.Coprod_mono Filter.coprodᵢ_mono
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**